File: src/providerState.ts
/** Opaque provider continuation data. Keep it out of tool arguments, chat text and progress events. */
export type ProviderState = {
  endpoint: string;
  model: string;
  outputItems: ProviderOutputItem[];
};

export type ProviderOutputItem =
  | {
    id: string;
    type: 'reasoning';
    summary: [];
    encrypted_content: string;
  }
  | {
    id?: string;
    type: 'function_call';
    call_id: string;
    name: string;
    arguments: string;
  }
  | {
    id?: string;
    type: 'message';
    role: 'assistant';
    content: { type: 'output_text'; text: string; annotations: [] }[];
    phase?: 'commentary' | 'final_answer';
  };

export const MAX_PROVIDER_STATE_CHARACTERS = 2_000_000;
export const MAX_PROVIDER_HISTORY_CHARACTERS = 8_000_000;

/** Validate data received from the backend or a saved checkpoint without interpreting encrypted reasoning. */
export function parseProviderState(value: unknown): ProviderState | undefined {
  if (!isRecord(value)
    || !hasOnlyKeys(value, ['endpoint', 'model', 'outputItems'])
    || !boundedString(value.endpoint, 2_048)
    || !boundedString(value.model, 120)
    || !Array.isArray(value.outputItems)
    || value.outputItems.length > 16) {
    return undefined;
  }
  const outputItems: ProviderOutputItem[] = [];
  for (const item of value.outputItems) {
    const parsed = parseOutputItem(item);
    if (!parsed) {
      return undefined;
    }
    outputItems.push(parsed);
  }
  const state: ProviderState = { endpoint: value.endpoint, model: value.model, outputItems };
  return JSON.stringify(state).length <= MAX_PROVIDER_STATE_CHARACTERS ? state : undefined;
}

function parseOutputItem(value: unknown): ProviderOutputItem | undefined {
  if (!isRecord(value) || (value.id !== undefined && !boundedString(value.id, 200))) {
    return undefined;
  }
  const identity = value.id === undefined ? {} : { id: value.id as string };
  if (value.type === 'reasoning'
    && hasOnlyKeys(value, ['id', 'type', 'summary', 'encrypted_content'])
    && boundedString(value.id, 200)
    && Array.isArray(value.summary) && value.summary.length === 0
    && boundedString(value.encrypted_content, 1_000_000)) {
    return { id: value.id, type: 'reasoning', summary: [], encrypted_content: value.encrypted_content };
  }
  if (value.type === 'function_call'
    && hasOnlyKeys(value, ['id', 'type', 'call_id', 'name', 'arguments'])
    && boundedString(value.call_id, 120)
    && boundedString(value.name, 120)
    && typeof value.arguments === 'string' && value.arguments.length <= 1_200_000) {
    return { ...identity, type: 'function_call', call_id: value.call_id, name: value.name, arguments: value.arguments };
  }
  if (value.type !== 'message'
    || !hasOnlyKeys(value, ['id', 'type', 'role', 'content', 'phase'])
    || value.role !== 'assistant'
    || !Array.isArray(value.content) || value.content.length > 8
    || (value.phase !== undefined && value.phase !== 'commentary' && value.phase !== 'final_answer')) {
    return undefined;
  }
  const content: Extract<ProviderOutputItem, { type: 'message' }>['content'] = [];
  for (const part of value.content) {
    if (!isRecord(part)
      || !hasOnlyKeys(part, ['type', 'text', 'annotations'])
      || part.type !== 'output_text'
      || typeof part.text !== 'string' || part.text.length > 400_000
      || !Array.isArray(part.annotations) || part.annotations.length !== 0) {
      return undefined;
    }
    content.push({ type: 'output_text', text: part.text, annotations: [] });
  }
  return { ...identity, type: 'message', role: 'assistant', content, ...(value.phase ? { phase: value.phase } : {}) };
}

function boundedString(value: unknown, maximum: number): value is string {
  return typeof value === 'string' && value.trim().length > 0 && value.length <= maximum;
}

function hasOnlyKeys(value: Record<string, unknown>, keys: string[]): boolean {
  return Object.keys(value).every((key) => keys.includes(key));
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

File: backend/app/provider_state.py
"""Validate the opaque Responses state kept between calls in one agent run."""

import json


MAX_PROVIDER_STATE_CHARACTERS = 2_000_000
MAX_PROVIDER_STATE_HISTORY_CHARACTERS = 8_000_000
MAX_PROVIDER_OUTPUT_ITEMS = 16
MAX_PROVIDER_CONTENT_PARTS = 8


def _text_size(value: str) -> int:
    try:
        return len(value.encode("utf-16-le")) // 2
    except UnicodeError:
        raise ValueError("providerState contains invalid text") from None


def serialized_provider_state_size(value: dict[str, object]) -> int:
    """Use the same compact JSON and UTF-16 units as the extension's state limit."""
    return _text_size(json.dumps(value, separators=(",", ":"), ensure_ascii=False))


def _record(value: object, required: set[str], optional: set[str] | None = None) -> dict:
    if not isinstance(value, dict) or not required <= value.keys():
        raise ValueError("providerState has an invalid structure")
    if not value.keys() <= required | (optional or set()):
        raise ValueError("providerState contains unsupported fields")
    return value


def _string(value: object, limit: int, *, allow_empty: bool = False) -> None:
    if not isinstance(value, str):
        raise ValueError("providerState has an invalid text field")
    if _text_size(value) > limit or (not allow_empty and not value.strip()):
        raise ValueError("providerState has an empty or oversized text field")


def _empty_list(value: object) -> None:
    if not isinstance(value, list) or value:
        raise ValueError("providerState summaries and annotations must be empty")


def _validate_output_item(value: object) -> None:
    if not isinstance(value, dict):
        raise ValueError("providerState has an invalid output item")
    kind = value.get("type")
    if kind == "reasoning":
        item = _record(value, {"id", "type", "summary", "encrypted_content"})
        _string(item["id"], 200)
        _empty_list(item["summary"])
        _string(item["encrypted_content"], 1_000_000)
    elif kind == "function_call":
        item = _record(value, {"type", "name", "call_id", "arguments"}, {"id"})
        _string(item["name"], 120)
        _string(item["call_id"], 120)
        _string(item["arguments"], 1_200_000, allow_empty=True)
    elif kind == "message":
        item = _record(value, {"type", "role", "content"}, {"id", "phase"})
        if item["role"] != "assistant":
            raise ValueError("providerState messages must have the assistant role")
        if "phase" in item and item["phase"] not in ("commentary", "final_answer"):
            raise ValueError("providerState has an invalid message phase")
        content = item["content"]
        if not isinstance(content, list) or len(content) > MAX_PROVIDER_CONTENT_PARTS:
            raise ValueError("providerState has too many message content parts")
        for raw_part in content:
            part = _record(raw_part, {"type", "text", "annotations"})
            if part["type"] != "output_text":
                raise ValueError("providerState contains an unsupported content type")
            _string(part["text"], 400_000, allow_empty=True)
            _empty_list(part["annotations"])
    else:
        raise ValueError("providerState contains an unsupported output type")
    if "id" in item:
        _string(item["id"], 200)


def validate_provider_state(value: object) -> dict[str, object]:
    """Check a small allowed shape without editing it or exposing its contents in errors.

    Reasoning is retained only as encrypted content. The provider adapter checks
    the endpoint, model, and matching function call before sending this state back.
    """
    state = _record(value, {"endpoint", "model", "outputItems"})
    _string(state["endpoint"], 2_048)
    _string(state["model"], 120)
    items = state["outputItems"]
    if not isinstance(items, list) or len(items) > MAX_PROVIDER_OUTPUT_ITEMS:
        raise ValueError("providerState has too many output items")
    for item in items:
        _validate_output_item(item)
    if serialized_provider_state_size(state) > MAX_PROVIDER_STATE_CHARACTERS:
        raise ValueError("providerState is too large")
    return state
